### DISCUSSION TITLE
Security - Design Authorization and Authentication System

### DISCUSSION THREAD

**Discussion by dmarro89** [dmarro89/dare-db#4] [Ideas]
Any idea on how to integrate authorization and authentication ?

**Comment by vdmitriyev** [comment]
There are different concepts and frameworks out there: [RBAC (Role-Based-Access-Control), ACL (Access-Control-List, ACLg (Access-Control-List grouped), etc.]. There are also a number of existing implementation in different databases and previous experience. However, I not quite sure, which one would be a good fit in this particular case. Thus, following should be done initially.

1. Research on best practices and approaches to secure users and roles in databases system design
2. Adapt found  concepts to the need to the current database
3. Discuss applicability of the concept to the current database
4. Implement the concept within the database
5. Integrate the concept into the database 
6. Write documentation on it

Links:

* [RBAC System](https://github.com/tarasowski/backend/blob/master/dynamodb/005_role_based_access_control.md)

**Reply by dmarro89** [reply]
I think we should start from a Role-based access control (RBAC).
Here's the steps:

- Define roles - Let's start with a simple implementation : only authenticated users can perform operations
- Define the permissions - read, write, delete 
- Setup authentication - Bearer Token - JWT ?
- Create endpoints and business logic for login and registration 
- Protect the endpoints with a middleware in order to guarantee the access to the features for the roles

**Reply by dmarro89** [reply]
I've created an issue -> https://github.com/dmarro89/dare-db/issues/7

**Reply by vdmitriyev** [reply]
I would also suggest to add here a concept of "schema". 

* In most of the databases, a new user will have its own schema assigned to him under the same name
* User would get "owner" right for the schema 
* The default user (e.g, admin) would have a different schema called **default**
* HTTP routing could be organized as follows
     - ```<URL-PREFIX>/<SCHEMA-NAME>/<PARAMETERS>```
* In case of the ```admin``` user the HTTP routing URL would look as follows (with ```default``` and without it - should be the same): 
     - ```<URL-PREFIX>/```**default**```/<PARAMETERS>```
     - ```<URL-PREFIX>/<PARAMETERS>```

**Comment by vdmitriyev** [comment]
Materials to consider for RBAC implementation:

* Building RBAC in Go - https://www.aserto.com/blog/building-rbac-in-go

**Comment by go-while** [comment]
related to https://github.com/dmarro89/dare-db/issues/7#issuecomment-2159096199

ssl already exists but did not try it yet

adding auth by ip / cidr can be added quickly, i have some copy and paste ready+

and code for webserver authentication via post/cookie, is copy paste too

**Reply by vdmitriyev** [reply]
Thank you for the comment. The concept should not be only about auth/etc. It should also take into consideration the underlying data, which may belong to different users/roles and users make have various roles as well. Thus, a user should not only be able to send/receive data, but in case of multiple users are using the same database, their data should be secured respectively.

Here is an draft of an architecture (see the RBAC part) for further details - https://github.com/dmarro89/dare-db/discussions/12 .

**Reply by go-while** [reply]
these are different layers of authentication. 

we should start with network level and work our way up but first get the core running before adding too much essentials. 

i'd be happy with ip and user/pass/cookie at the moment.

that's an awesome idea and i don't see a problem in writing it :long_drum:

**Comment by vdmitriyev** [comment]
For now the issue related RBAC was closed and moved into backlog: https://github.com/dmarro89/dare-db/issues/7

A new issue on basic auth has been created: https://github.com/dmarro89/dare-db/issues/33

**Comment by dmarro89** [comment]
Hi @vdmitriyev, how are you ?
Have you been able to see the PR -> https://github.com/dmarro89/dare-db/pull/35?

Cheers

**Reply by vdmitriyev** [reply]
@dmarro89 thank you for this PR. I will try to to review your changes next week. I hope, you are fine with that.

**Reply by dmarro89** [reply]
Yes of course!
Thank you very much

**Comment by dmarro89** [comment]
@vdmitriyev Hi,
about the Authorization and Authentication, I think that there are few things missing:

- Endpoints to Create New users -> only for Admins ROLES
- Endpoints to update password for all users
- Endpoints to update Authorization for each users -> only for Admins

What do you think? Am I missing something ?

@TheInvincibleRalph
